src/queue.py

```diff
@@ -0,0 +1,59 @@
+"""Implementation of queue data structure."""
+
+
+class Queue(object):
+    """A singly-linked list."""
+
+    def __init__(self):
+        """Create an instance of type Queue. Allow data to be passed in."""
+        self.head = None
+        self.tail = None
+
+    def enqueue(self, val):
+        """Append a val to the end of the queue."""
+        node = Node(val, None, self.tail)
+        if self.size() == 0:
+            self.head = node
+        old_tail = self.tail
+        if self.tail is not None:
+            self.tail.next = node
+        self.tail = node
+        self.tail.prev = old_tail
+
+    def dequeue(self):
+        """Remove and returns the first item in the queue."""
+        try:
+            val = self.head.data
+            if self.head.next is None:
+                self.tail = None
+            self.head.prev = None
+            self.head = self.head.next
+            return val
+        except AttributeError:
+            raise AttributeError("Cannot pop from an empty list.")
+
+    def size(self):
+        """Count the number of items in the queue."""
+        cur = self.head
+        count = 0
+        if self.head:
+            count = 1
+        while self.tail is not cur:
+            count += 1
+            cur = cur.next
+        return count
```

Comment:
Why iterate through the whole thing when you could keep track of the size on each enqueue and dequeue?